### PR TITLE
Soft tigger STXM scan

### LINKED ISSUE #34: Hit and hope fly scan.
### Goals
One of the three scanning type needed for STXM.
Acceptance criteria : 
- [x] Flying scan where the motor moves to the starting point and start moving to toward the end point, during which the detector keep measuring as fast as it can until the endpoint is reached or stopped. 
- [x] Each measurement should matches with the motor position as closely as possible but do not have to be exact like step 
- [x] User should be able to define the two axis they want to scan, time they want it to take to cover the area and  count time on the detector, the software then calculate the speed needed to cover the scanning area and does the fast-STXM scan.  

### Tasks
- [x] Create 1d fly scan. 
- [x] Add an axis so, the user can do multiple lines of fast scan within one plan.
- [x] Add speed calculation  to allow user to do scan with time instead of speed. 
     1. Error checking, need to check the scan can actually be done, due to max/min speed restriction.

### PR DESCRIPTION
**Pull request by Relm-Arrowny**
Fix #34 

Added stxm_fast plans, see #34 for detail description. 

### Instructions to reviewer on how to test:
1. Run unit tests.
2. Please ignore jupyter_tests folder, it is mostly sandbox/record keeping for me.
3. It should work but I am mostly worry about correctness and best practice any feedback would be much appreciated. 

### Checks for reviewer
- [ ] Would the PR title make sense to a scientist on a set of release notes
- [ ] If a new device has been added does it follow the [standards](https://github.com/DiamondLightSource/dodal/wiki/Device-Standards)

### CODE REVIEW

**Comment by Relm-Arrowny** [comment]
This is far from ready but am worry that this is not the right way to go about it, in pacifically the logic of this function:[_fast_scan_1d](https://github.com/DiamondLightSource/p99-bluesky/blob/75176f43e637aa392eec9bb345d9e1f48a76bdc9/src/p99_bluesky/plans/fast_scan.py#L140).  I am wondering if you have time could you have look and make sure I have not missed a better way of doing it.  @DominicOram @callumforrester @stan-dot

File: src/p99_bluesky/plans/fast_scan.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+    yield from finalize_wrapper(
+        plan=inner_fast_scan_1d(dets, motor, start, end, motor_speed),
+        final_plan=reset_speed(old_speed, motor),
+    )
+
+
+def check_within_limit(values: list, motor: Motor):
+    LOGGER.info(f"Check {motor.name} limits.")
+    lower_limit = yield from bps.rd(motor.low_limit_travel)
+    high_limit = yield from bps.rd(motor.high_limit_travel)
```
Comment (stan-dot) [inline src/p99_bluesky/plans/fast_scan.py:214]: perhaps move this to `p99_bluesky/src/stubs/check_motor_within_limit`?  with the view of moving into dodal later?

File: src/p99_bluesky/plans/fast_scan.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+        step_motor: Motor,
+        step_start: float,
+        step_end: float,
+        step_number: float,
+        scan_motor: Motor,
+        scan_start: float,
+        scan_end: float,
+        motor_speed: float | None = None,
+        snake_axes: bool = False,
+    ):
```
Comment (stan-dot) [inline src/p99_bluesky/plans/fast_scan.py:98]: it's missing the `-> MsgGenerator` I think

File: src/p99_bluesky/plans/fast_scan.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+                )
+                step_counter += 1
+                yield from bps.mv(step_motor, step_start + step_size * step_counter)
+                yield from _fast_scan_1d(
+                    dets + [step_motor], scan_motor, scan_end, scan_start, motor_speed
+                )
+                step_counter += 1
+        else:
+            while step_number >= step_counter:
+                yield from bps.mv(step_motor, step_start + step_size * step_counter)
```
Comment (stan-dot) [inline src/p99_bluesky/plans/fast_scan.py:117]: nit: extract the target location from the move line

File: src/p99_bluesky/plans/stxm.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    scan_motor: Motor,
+    scan_start: float,
+    scan_end: float,
+    plan_time: float,
+    # step_size: float,
+):
+    num_data_point = plan_time / count_time
+    scan_range = abs(scan_start - scan_end)
+    step_range = abs(step_start - step_end)
+    ideal_step_size = 1.0 / ((num_data_point / (scan_range * step_range)) ** 0.5)
```
Comment (stan-dot) [inline src/p99_bluesky/plans/stxm.py:46]: should: add a comment why this is the ideal step size

File: src/p99_bluesky/plans/stxm.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    ideal_velocity = ideal_step_size / count_time
+    LOGGER.info(f"{ideal_step_size} velocity = {ideal_velocity}.")
+    velocity, step_size = yield from get_velocity_and_step_size(
+        scan_motor, ideal_velocity, ideal_step_size
+    )
+    LOGGER.info(f"{scan_motor.name} velocity = {velocity}.")
+    LOGGER.info(f"{step_motor.name} step size = {step_size}.")
+    # yield from bps.abs_set(det.drv.acquire_time, count_time)
+    import math
```
Comment (stan-dot) [inline src/p99_bluesky/plans/stxm.py:55]: not sure why import mid-function

File: src/p99_bluesky/plans/stxm.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+#     step_motor: Motor,
+#     step_start: float,
+#     step_end: float,
+#     step_size: float,
+#     scan_motor: Motor,
+#     scan_start: float,
+#     scan_end: float,
+#     motor_speed: float | None = None,
+#     snake_axes: bool = False,
+# ):
```
Comment (stan-dot) [inline src/p99_bluesky/plans/stxm.py:104]: delete the commented out code

**Comment by Relm-Arrowny** [comment]
[Flyable motor may be better way of doing this.](https://github.com/bluesky/ophyd-async/pull/425),

File: tests/test_fast_scan.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+
+# async def test_fast_scan_grid_success(
+#     sim_motor: ThreeAxisStage, RE: RunEngine, andor2: Andor2Ad
+# ):
+#     RE(fast_scan_grid([det], sim_motor.x, 5, -1, 10, sim_motor.y, 5, -1, 10))
+
+#     assert 2.78 == await sim_motor.y.velocity.get_value()
+#     assert 20 == get_mock_put(sim_motor.y.user_setpoint).call_count
+#     assert 20 == get_mock_put(sim_motor.y.velocity).call_count
```
Comment (stan-dot) [inline tests/test_fast_scan.py:100]: should: delete commented out code

**Review by stan-dot — CHANGES_REQUESTED** [review]

File: pyproject.toml [hunk 8]
```diff
@@ -123,7 +123,7 @@ allowlist_externals =
 commands =
     pre-commit: pre-commit run --all-files {posargs}
     type-checking: pyright src tests {posargs}
-    tests: pytest --cov=p99_bluesky --cov-report term --cov-report xml:cov.xml {posargs}
+    tests: pytest --cov=p99_bluesky --cov-report term  --cov-report xml:cov.xml {posargs}
```
[the remark this answers is not in the document]
Comment (callumforrester) [reply pyproject.toml]: It's still there

**Comment by callumforrester** [comment]
@Relm-Arrowny this may be for a later PR but you should consider starting to use [scanspec](https://github.com/dls-controls/scanspec)

**Pull request merged**